crates/tarball/src/lib.rs

```diff
@@ -194,36 +209,75 @@ impl<'a> DownloadTarballToStore<'a> {
-                    entry.read_to_end(&mut buffer).unwrap();
-
-                    let entry_path = entry.path().unwrap();
-                    let cleaned_entry_path =
-                        entry_path.components().skip(1).collect::<PathBuf>().into_os_string();
-                    let integrity = pacquet_cafs::write_sync(store_dir, &buffer)
-                        .map_err(TarballError::WriteCafs)?;
-
-                    Ok((cleaned_entry_path, store_dir.join(integrity)))
-                })
-                .collect::<Result<HashMap<OsString, PathBuf>, TarballError>>()
-                .map_err(TaskError::Other)
+                .filter(|entry| !entry.as_ref().unwrap().header().entry_type().is_dir());
+
+            let ((_, Some(capacity)) | (capacity, None)) = entries.size_hint();
+            let mut cas_paths = HashMap::<OsString, PathBuf>::with_capacity(capacity);
+            let mut tarball_index = TarballIndex { files: HashMap::with_capacity(capacity) };
+
+            for entry in entries {
+                let mut entry = entry.unwrap();
+
+                let file_mode = entry.header().mode().expect("get mode"); // TODO: properly propagate this error
+                let is_executable = file_mode & EXEC_MASK != 0;
+
+                // Read the contents of the entry
+                let mut buffer = Vec::with_capacity(entry.size() as usize);
+                entry.read_to_end(&mut buffer).unwrap();
+
+                let entry_path = entry.path().unwrap();
+                let cleaned_entry_path =
+                    entry_path.components().skip(1).collect::<PathBuf>().into_os_string();
+                let (file_path, file_hash) = store_dir
+                    .write_cas_file(&buffer, is_executable)
+                    .map_err(TarballError::WriteCasFile)?;
+
+                let tarball_index_key = cleaned_entry_path
+                    .to_str()
+                    .expect("entry path must be valid UTF-8") // TODO: propagate this error, provide more information
+                    .to_string(); // TODO: convert cleaned_entry_path to String too.
+
+                if let Some(previous) = cas_paths.insert(cleaned_entry_path, file_path) {
+                    panic!("Unexpected error: {previous:?} shouldn't collide");
+                }
+
+                let checked_at = UNIX_EPOCH.elapsed().ok().map(|x| x.as_millis());
+                let file_size = entry.header().size().ok();
+                let file_integrity = format!("sha512-{}", BASE64_STD.encode(file_hash));
+                let file_attrs = TarballIndexFileAttrs {
+                    checked_at,
+                    integrity: file_integrity,
+                    mode: file_mode,
+                    size: file_size,
+                };
+
+                if let Some(previous) = tarball_index.files.insert(tarball_index_key, file_attrs) {
+                    panic!("Unexpected error: {previous:?} shouldn't collide");
+                }
+            }
```

Comment:
Same as above. The last file should be used.

Reply:
https://github.com/pnpm/pacquet/pull/166/commits/3ab34b855644f2ed22ecb87a35aa8a87922f1b9c